File: Source/Entities/Entities/Ch15/FlashingBlack.cs
namespace Celeste.Entities.Chapters.Ch15
{
    /// <summary>
    /// FlashingBlack - Citadel blackout. Cycles visible, a warning flash, then a full
    /// black screen for <c>blackDuration</c> before fading back. The room stays fully
    /// playable in the dark, so it is a memory test rather than a hazard.
    ///
    /// One instance covers the whole room; it renders over everything via a very
    /// negative depth. <c>keepPlayerVisible</c> leaves a faint marker at the player so
    /// blackouts can be tuned from "hard" to "brutal".
    ///
    /// Placeholder art: full-screen black square (a small marker square in the editor).
    /// </summary>
    [CustomEntity("DZ/FlashingBlack")]
    [Tracked]
    public class FlashingBlack : Entity
    {
        #region Enums
        private enum Phase
        {
            Visible,
            FadingOut,
            Black,
            FadingIn
        }
        #endregion

        #region Fields
        private readonly float visibleDuration;
        private readonly float fadeDuration;
        private readonly float blackDuration;
        private readonly float maxAlpha;
        private readonly bool keepPlayerVisible;
        private readonly bool warnFlash;
        private readonly string flag;
        private readonly bool invertFlag;

        private Phase phase = Phase.Visible;
        private float phaseTimer;
        private float alpha;
        private float warnAlpha;
        private Level level;
        #endregion

        #region Constructor
        public FlashingBlack(EntityData data, Vector2 offset)
            : base(data.Position + offset)
        {
            visibleDuration = Math.Max(0.1f, data.Float("visibleDuration", 2.4f));
            fadeDuration = Math.Max(0.01f, data.Float("fadeDuration", 0.35f));
            blackDuration = Math.Max(0.1f, data.Float("blackDuration", 1.1f));
            maxAlpha = MathHelper.Clamp(data.Float("maxAlpha", 1f), 0f, 1f);
            keepPlayerVisible = data.Bool("keepPlayerVisible", false);
            warnFlash = data.Bool("warnFlash", true);
            flag = data.Attr("flag", "");
            invertFlag = data.Bool("invertFlag", false);

            if (data.Bool("startBlack", false))
            {
                phase = Phase.Black;
                alpha = maxAlpha;
                phaseTimer = blackDuration;
            }
            else
            {
                phaseTimer = visibleDuration;
            }

            // Above every normal entity and the player.
            Depth = -1000000;
        }
        #endregion

        #region Private
        /// <summary>Whether this entity's enable flag currently permits it to run.</summary>
        private bool FlagAllows()
        {
            if (string.IsNullOrEmpty(flag) || level == null)
            {
                return true;
            }

            return level.Session.GetFlag(flag) != invertFlag;
        }
        #endregion

        #region Entity Overrides
        public override void Added(Scene scene)
        {
            base.Added(scene);
            level = scene as Level;
        }

        public override void Update()
        {
            base.Update();

            warnAlpha = Calc.Approach(warnAlpha, 0f, Engine.DeltaTime * 4f);

            if (!FlagAllows())
            {
                alpha = Calc.Approach(alpha, 0f, Engine.DeltaTime / fadeDuration);
                return;
            }

            phaseTimer -= Engine.DeltaTime;

            switch (phase)
            {
                case Phase.Visible:
                    alpha = Calc.Approach(alpha, 0f, maxAlpha / fadeDuration * Engine.DeltaTime);

                    if (warnFlash && phaseTimer <= 0.25f && warnAlpha <= 0f && phaseTimer > 0f)
                    {
                        warnAlpha = 0.5f;
                        Audio.Play("event:/game/general/thing_booped", Position);
                    }

                    if (phaseTimer <= 0f)
                    {
                        phase = Phase.FadingOut;
                        phaseTimer = fadeDuration;
                    }
                    break;

                case Phase.FadingOut:
                    alpha = Calc.Approach(alpha, maxAlpha, maxAlpha / fadeDuration * Engine.DeltaTime);

                    if (phaseTimer <= 0f)
                    {
                        phase = Phase.Black;
                        phaseTimer = blackDuration;
                        alpha = maxAlpha;
                    }
                    break;

                case Phase.Black:
                    if (phaseTimer <= 0f)
                    {
                        phase = Phase.FadingIn;
                        phaseTimer = fadeDuration;
                    }
                    break;

                case Phase.FadingIn:
                    alpha = Calc.Approach(alpha, 0f, maxAlpha / fadeDuration * Engine.DeltaTime);

                    if (phaseTimer <= 0f)
                    {
                        phase = Phase.Visible;
                        phaseTimer = visibleDuration;
                        alpha = 0f;
                    }
                    break;
            }
        }

        public override void Render()
        {
            base.Render();

            if (level == null || (alpha <= 0.002f && warnAlpha <= 0.002f))
            {
                return;
            }

            // Cover the camera view with a margin so no edge shows during shake.
            Vector2 topLeft = level.Camera.Position - new Vector2(16f, 16f);
            const float width = 320f + 32f;
            const float height = 180f + 32f;

            if (warnAlpha > 0.002f)
            {
                Draw.Rect(topLeft.X, topLeft.Y, width, height, Color.White * warnAlpha * 0.35f);
            }

            if (alpha > 0.002f)
            {
                Draw.Rect(topLeft.X, topLeft.Y, width, height, Color.Black * alpha);

                if (keepPlayerVisible)
                {
                    Player player = Scene.Tracker.GetEntity<Player>();
                    if (player != null)
                    {
                        Placeholder.FillCircle(player.Center, 4f, Color.White * (alpha * 0.35f));
                    }
                }
            }
        }

        public override void DebugRender(Camera camera)
        {
            base.DebugRender(camera);
            Draw.HollowRect(X - 8f, Y - 8f, 16f, 16f, Color.Gray);
        }
        #endregion
    }
}

File: Loenn/entities/flashingBlack.lua
-- Loenn plugin for DZ/FlashingBlack (chapter 15 - Citadel).
-- Placeholder visual: small black marker square (the effect is full-screen in game). Swap
-- the sprite function for a drawableSprite once real art exists; no placement data has to
-- change.

local drawableRectangle = require("structs.drawable_rectangle")
local utils = require("utils")

local flashingBlack = {}

flashingBlack.name = "DZ/FlashingBlack"
flashingBlack.depth = -1000000

flashingBlack.placements = {
    name = "flashing_black",
    data = {
        visibleDuration = 2.4,
        fadeDuration = 0.35,
        blackDuration = 1.1,
        maxAlpha = 1.0,
        keepPlayerVisible = false,
        warnFlash = true,
        startBlack = false,
        flag = "",
        invertFlag = false
    }
}

flashingBlack.fieldInformation = {
    visibleDuration = {
        minimumValue = 0.1
    },
    fadeDuration = {
        minimumValue = 0.01
    },
    blackDuration = {
        minimumValue = 0.1
    },
    maxAlpha = {
        minimumValue = 0.0,
        maximumValue = 1.0
    }
}

-- drawableRectangle returns a single sprite in "fill" mode and a list of sprites in
-- "line" mode, so flatten whatever comes back before handing it to Loenn.
local function push(sprites, rectangle)
    local result = rectangle:getDrawableSprite()

    if result[1] ~= nil then
        for _, sprite in ipairs(result) do
            table.insert(sprites, sprite)
        end
    else
        table.insert(sprites, result)
    end
end

local function pushRect(sprites, x, y, width, height, fill, line)
    push(sprites, drawableRectangle.fromRectangle("fill", x, y, width, height, fill))
    push(sprites, drawableRectangle.fromRectangle("line", x, y, width, height, line))
end

local fillColor = {0.05, 0.05, 0.07, 0.95}
local lineColor = {0.65, 0.65, 0.7, 1.0}

-- One instance blacks out the whole room, so the editor only needs a marker.
local markerSize = 16

function flashingBlack.sprite(room, entity)
    local x, y = (entity.x or 0) - markerSize / 2, (entity.y or 0) - markerSize / 2
    local sprites = {}

    pushRect(sprites, x, y, markerSize, markerSize, fillColor, lineColor)

    return sprites
end

function flashingBlack.selection(room, entity)
    return utils.rectangle((entity.x or 0) - markerSize / 2, (entity.y or 0) - markerSize / 2,
        markerSize, markerSize)
end

return flashingBlack
